mail-server/borgbackup.nix

```diff
@@ -0,0 +1,79 @@
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program. If not, see <http://www.gnu.org/licenses/>
+
+{ config, pkgs, lib, ... }:
+
+let
+  cfg = config.mailserver.borgbackup;
+
+  methodFragment = lib.optional (cfg.compression.method != null) cfg.compression.method;  
+  autoFragment =
+    if cfg.compression.auto && cfg.compression.method == null
+    then throw "compression.method must be set when using auto."
+    else lib.optional cfg.compression.auto "auto";
+  levelFragment =
+    if cfg.compression.level != null && cfg.compression.method == null
+    then throw "compression.method must be set when using compression.level."
+    else lib.optional (cfg.compression.level != null) (toString cfg.compression.level);
+  compressionFragment = lib.concatStringsSep "," (lib.flatten [autoFragment methodFragment levelFragment]);
+  compression = lib.optionalString (compressionFragment != "") "--compression ${compressionFragment}";
+
+  encryptionFragment = cfg.encryption.method;
+  passphraseFile = lib.escapeShellArg cfg.encryption.passphraseFile;
+  passphraseFragment = lib.optionalString (cfg.encryption.method != "none")
+                         (if cfg.encryption.passphraseFile != null then ''env BORG_PASSPHRASE="$(cat ${passphraseFile})"''
+                          else throw "passphraseFile must be set when using encryption.");
+
+  locations = lib.escapeShellArgs cfg.locations;
+  name = lib.escapeShellArg cfg.name;
+
+  repoLocation = lib.escapeShellArg cfg.repoLocation;
+
+  extraInitArgs = lib.escapeShellArgs cfg.extraArgumentsForInit;
+  extraCreateArgs = lib.escapeShellArgs cfg.extraArgumentsForCreate;
+
+  cmdPreexec = lib.optionalString (cfg.cmdPreexec != null) cfg.cmdPreexec;
+  cmdPostexec = lib.optionalString (cfg.cmdPostexec != null) cfg.cmdPostexec;
+
+  borgScript = ''
+    . /etc/profile
+    export BORG_REPO=${repoLocation}
+    ${cmdPreexec}
+    ${passphraseFragment} ${pkgs.borgbackup}/bin/borg init ${extraInitArgs} --encryption ${encryptionFragment} || true
+    ionice -c 3 ${passphraseFragment} ${pkgs.borgbackup}/bin/borg create ${extraCreateArgs} ${compression} ::${name} ${locations}
+    ${cmdPostexec}
+  '';
+in {
+  config = lib.mkIf config.mailserver.borgbackup.enable {
+    environment.systemPackages = with pkgs; [
+      borgbackup
+    ];
+
+    systemd.services.borgbackup = {
+      description = "borgbackup";
+      unitConfig.Documentation = "man:borgbackup";
+      script = borgScript;
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        CPUSchedulingPolicy = "idle";
+        IOSchedulingClass = "idle";
+        ProtectSystem = "full";
```

Comment:
Why not
```
ProtectSystem = "strict";
ReadWritePaths = lib.optional (builtins.substring 0 1 cfg.repoLocation == "/") cfg.repoLocation;
```
where the thing inside paratheses is something like `isPath`.

Reply:
I did not intend to spend as much time as I did on this in the first place and adding `ProtectSystem = "full"` was a rather low hanging fruit that needed no further thinking.

Using `strict` already requires more nix-fu and borg does actually create files in places other than `repoLocation` as well: 

- `~/.cache/borg` and
- `~/.config/borg`.

In addition to that the important bits on a NixOS system are already mounted read-only.

Reply:
Alright, I'll look into this a little more when I write a NixOS module (probably next week).